designs/2018-programmatic-style-edits/README.md

```diff
@@ -0,0 +1,401 @@
+
+All of this is to say that the current way that ESLint deals with style enforcers, including its interaction with Prettier, leaves a lot of room for improvement.
+
+### Goals
+
+The goals of this programmatic style editor feature are:
+
+1. Allow style enforcers that do not rely on rules.
+1. Allow arbitrary modification of source code within the context of autofixing.
+1. Enable the use of third-party tools to do this modification easily.
+1. Reduce developers' dependence on ESLint core rules for style guide enforcement and therefore reduce the maintenance burden on the ESLint team.
+
+## Detailed Design
+
+At a high-level, this design entails:
+
+1. Enable plugins to export style editors
+1. Allow third-party tools to be incorporated into style editors
+1. Automatically apply style editors with using `--fix`
+1. Allow turning on/off style editors using `--fix-type`
+1. Ensure style editors are composable
+
+Definitions:
+
+* **Style Edit** - an individual code modification written in JavaScript.
+* **Style Editor** - a collection of style edits to apply in order. Style editors are applied through configuration.
+
+### Style Editors in Plugins
+
+A plugin can contain one or more style editors, exported on a `styles` object in a manner similar to `rules` and `processors`, for example:
+
+```js
+// eslint-plugin-mystyles
+const es5StyleEditor = require("./styles/es5");
+
+module.exports = {
+    styles: {
+        es5: es5StyleEditor
+    }
+};
+```
+
+In this code, the plugin `eslint-plugin-mystyles` exports a style editor called `es5`. In order to apply the style editor to code, the end-user would add a `style` property to the configuration object:
+
+```js
+module.exports = {
+    plugins: ["mystyles"],
+    parserOptions: {
+        ecmaVersion: 5
+    },
+    style: "mystyles/es5"
+}
+```
+
+The `style` property can be used at the top level of the configuration to apply to all files or in an `overrides` object to apply to a subset of files.
+
+Options for style editors can be specified using a `styleOptions` key, as in this example:
+
+```js
+module.exports = {
+    plugins: ["mystyles"],
+    parserOptions: {
+        ecmaVersion: 5
+    },
+    style: "mystyles/es5",
+    styleOptions: {
+        indent: "tabs"
+    }
+}
+```
+
+The `styleOptions` property is just an opaque object to ESLint, without any predefined properties. Style editors can define their options and ESLint will pass the `styleOptions` object directly into the style editor through `context.options`.
+
+### Style Editor Format
+
+A style editor is an object literal with a `meta` property to provide information about the style editor and an `edits` property that is an array of the style edits to apply, for example:
+
+```js
+// style editor format
+module.exports = {
+    meta: {
+        name: "ES5 Style",
+        description: "A style editor just for ES5"
+    },
+    edits: [
+        // individual style edits
```

Comment:
Are there use cases where the entries in the `edits` array would vary based on `styleOptions`? I'm wondering if it's more future-proof to make this similar to the rule API where we have `meta` and a function to return the `edits` array (whether that function is named `create` like in the rule API or still `edits` and we just check `typeof styleEditor.edits === "function"` before calling it or using its array value).

Everything I can think of could be done at the individual edit level, where options are already available in the current iteration of the RFC, but I'm curious whether you have ideas for future directions that might need the additional flexibility.